cmd/client/txgen/main.go

```diff
@@ -89,166 +95,88 @@ func main() {
-					utils.GetLogInstance().Info("Adding block from leader", "txNum", len(block.Transactions()), "shardID", shardID, "preHash", block.ParentHash().Hex())
-					node.AddNewBlock(block)
-					stateMutex.Lock()
-					node.Worker.UpdateCurrent()
-					stateMutex.Unlock()
-					readySignal <- shardID
-				} else {
-					continue
-				}
-			}
-		}
-	}
-	clientNode.Client.UpdateBlocks = updateBlocksFunc
-
-	clientNode.NodeConfig.SetRole(nodeconfig.ClientNode)
-	clientNode.NodeConfig.SetIsClient(true)
-	clientNode.ServiceManagerSetup()
-	clientNode.RunServices()
-
-	// Start the client server to listen to leader's message
-	go func() {
-		// wait for 3 seconds for client to send ping message to leader
-		// FIXME (leo) the readySignal should be set once we really sent ping message to leader
-		time.Sleep(3 * time.Second) // wait for nodes to be ready
-		for _, i := range shardIDs {
-			readySignal <- i
-		}
-	}()
-
-	// Transaction generation process
+	txGen := node.New(myhost, consensusObj, nil, false) //Changed it : no longer archival node.
+	txGen.Client = client.NewClient(txGen.GetHost(), shardIDs)
+	txGen.NodeConfig.SetRole(nodeconfig.ClientNode)
+	txGen.NodeConfig.SetIsBeacon(true)
+	txGen.NodeConfig.SetIsClient(true)
+	txGen.NodeConfig.SetShardGroupID(p2p.GroupIDBeacon)
+	txGen.ServiceManagerSetup()
+	txGen.RunServices()
+	time.Sleep(20 * time.Second)
+	utils.GetLogInstance().Debug("node nighbors", "neig", txGen.BeaconNeighbors, "normal", txGen.Neighbors)
 	start := time.Now()
 	totalTime := float64(*duration)
-
+	ticker := time.NewTicker(checkFrequency * time.Second)
+	txGen.GetSync()
 	for {
 		t := time.Now()
 		if totalTime > 0 && t.Sub(start).Seconds() >= totalTime {
 			utils.GetLogInstance().Debug("Generator timer ended.", "duration", (int(t.Sub(start))), "startTime", start, "totalTime", totalTime)
 			break
 		}
 		select {
-		case shardID := <-readySignal:
-			shardIDTxsMap := make(map[uint32]types.Transactions)
-			lock := sync.Mutex{}
-
-			stateMutex.Lock()
-			utils.GetLogInstance().Warn("STARTING TX GEN", "gomaxprocs", runtime.GOMAXPROCS(0))
-			txs, _ := txgen.GenerateSimulatedTransactionsAccount(int(shardID), nodes, setting)
-
-			lock.Lock()
-			// Put txs into corresponding shards
-			shardIDTxsMap[shardID] = append(shardIDTxsMap[shardID], txs...)
-			lock.Unlock()
-			stateMutex.Unlock()
-
-			lock.Lock()
-			for shardID, txs := range shardIDTxsMap { // Send the txs to corresponding shards
-				go func(shardID uint32, txs types.Transactions) {
-					SendTxsToShard(clientNode, txs)
-				}(shardID, txs)
+		case <-ticker.C:
+			if txGen.State.String() == "NodeReadyForConsensus" {
+				utils.GetLogInstance().Debug("Generator Will Send Txns.", "txgen node", txGen.SelfPeer, "Node State", txGen.State.String())
+				txs, _ := GenerateSimulatedTransactionsAccount(int(shardID), txGen, setting)
+				SendTxsToShard(txGen, txs)
+				go txGen.GetSync()
```

Comment:
why do sync for every block?

Reply:
You mean why do sync for every batch? We need to be in sync to be able to generate valid txns? Not sure I get your question.

Reply:
For a client node, it should automatically receive new blocks from the network. State sync'ing should only happen at start. State sync'ing is used to catch up, not to stay in sync.

Reply:
Resolved now by receiving new blocks from network.